src/quartics/tenten_qs.py

```diff
@@ -1,2 +1,656 @@
+            if realcase[0] == -1 or err1 < err0:
+                whichcase = 1 # d2 = 0
+                if realcase[1] == 1:
+                    aq = aq1
+                    bq = bq1
+                    cq = cq1
+                    dq = dq1
+
+                else:
+                    acx = acx1
+                    bcx = bcx1
+                    ccx = ccx1
+                    dcx = dcx1
+        if realcase[whichcase] == 1:
+            # If alpha1, beta1, alpha2, and beta2 are real first refine them through a Newton-Ralphson
+            aq, bq, cq, dq = self._newton_raphson([a, b, c, d], [aq, bq, cq, dq])
+
+            # Finally calculate roots as roots of p1(x) and p2(x) (end of section 2.1)
+            qroots = self._solve_quadratic(aq, bq, [None, None])
+            final_roots[0:2] = qroots
+            qroots = self._solve_quadratic(cq, dq, qroots)
+            final_roots[2:4] = qroots
+
+        else:
+            # Complex coefficients of p1 and p2
+            if whichcase == 0: # d2 != 0 
+                cdiskr = 0.25*sq(acx) - bcx
+                # Calculate roots as those of p1(x) and p2(x) (end of sec. 2.1)
+                zx1 = -0.5*acx + sqrt(cdiskr)
+                zx2 = -0.5*acx - sqrt(cdiskr)
+                zxmax = zx1 if abs(zx1) > abs(zx2) else zx2
+                zxmin = bcx / zxmax
+                final_roots[0:4] = [
+                    zxmin,
+                    mpmath.conj(zxmin),
+                    zxmax,
+                    mpmath.conj(zxmax)
+                ]
+            else: # d2 ~= 0
+                # Theoretically this path should never be reached
+                cdiskr = sqrt(sq(acx) - 4*bcx)
+                zx1 = -0.5 * (acx + cdiskr)
+                zx2 = -0.5 * (acx - cdiskr)
+                zxmax = zx1 if abs(zx1) > abs(zx2) else zx2
+                zxmin = dcx / zxmax
+                final_roots[0:2] = [
+                    zxmax,
+                    zxmin
+                ]
+                cdiskr = sqrt(sq(ccx) - 4*dcx)
+                zx1 = -0.5 * (ccx + cdiskr)
+                zx2 = -0.5 * (ccx - cdiskr)
+                zxmax = zx1 if abs(zx1) > abs(zx2) else zx2
+                zxmin = dcx / zxmax
+                final_roots[2:4] = [
+                    zxmax,
+                    zxmin
+                ]
+        if rfact != mpf(1):
+            for k in range(4):
+                final_roots[k] *= rfact
+        return final_roots
+
+
+def copysign(sign_of: MpfAble, magn_of: MpfAble) -> mpf:
+    '''
+    Mimic std::copysign / math.copysign but make sure to keep it in mpf
+    '''
+    return mpf(math.copysign(1, mpf(sign_of))) * mpf(magn_of)
+
+
+def sq(val: mpf):
+    '''
+    Shorthand for the square of a number, for readability
+    '''
+    return val*val
+
+
+def cbrt(val: mpc):
+    '''
+    Return cube root while maintaining sign
+    '''
+    if val.real >= 0:
+        return mpmath.cbrt(val)
+    else:
+        return -mpmath.cbrt(-val)
```

Comment:
rename this "real_cube_root", or something that better communicates that you are specifically obtain the real root.